compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeNpuDmaCpyNd.cpp

```diff
@@ -0,0 +1,194 @@
+          dmaOp.emitOpError()
+              << "has target in L3, but does not have target addressing. "
+                 "Target addressing is required to canonicalize here.";
+          return WalkResult::interrupt();
+        }
+        tgtOffsets = getPrepended(tgtOffsets, zero);
+        tgtSizes = getPrepended(tgtSizes, one);
+        tgtStrides = getPrepended(tgtStrides, zero);
+        std::optional<uint32_t> maybeSwapIndex =
+            verifyAndGetZeroStrideIndex(tgtSizes, tgtStrides, dmaOp);
+        if (!maybeSwapIndex.has_value()) {
+          return WalkResult::interrupt();
+        }
+        uint32_t swapIndex = maybeSwapIndex.value();
+        bubble(tgtOffsets, swapIndex);
+        bubble(tgtSizes, swapIndex);
+        bubble(tgtStrides, swapIndex);
+      }
+
+      rewriter.setInsertionPoint(dmaOp);
+
+      // Replace the npu.dma_cpy_nd with the canonicalized version.
+      dmaOp = rewriter.replaceOpWithNewOp<AMDAIE::NpuDmaCpyNdOp>(
+          dmaOp, dmaOp.getDma(), dmaOp.getTarget(), tgtOffsets, tgtSizes,
+          tgtStrides, dmaOp.getTargetBdId(), dmaOp.getSource(), srcOffsets,
+          srcSizes, srcStrides, dmaOp.getSourceBdId());
+
+      return WalkResult::advance();
+    });
+
+    if (walkResult.wasInterrupted()) {
+      return signalPassFailure();
+    }
+  }
+
+ private:
+  /// Repeat prepend 'def' to 'tail' to make 'tail' have nbDimensions elements.
+  SmallVector<OpFoldResult> getPrepended(ArrayRef<OpFoldResult> tail,
+                                         Attribute def) {
+    assert(tail.size() <= nbDimensions);
+    SmallVector<OpFoldResult> res(nbDimensions, def);
+    std::copy(tail.begin(), tail.end(),
+              res.begin() + nbDimensions - tail.size());
+    return res;
+  }
+
+  static size_t getLowestIndexMaybeAboveOne(ArrayRef<OpFoldResult> v) {
+    for (size_t i = 0; i < v.size(); i++) {
+      std::optional<int64_t> maybe = getConstantIntValue(v[i]);
+      if (!maybe.has_value() || maybe.value() > 1) {
+        return i;
+      }
+    }
+    return v.size();
+  }
+
+  static size_t getHighestIndexMaybeZero(ArrayRef<OpFoldResult> v) {
+    for (size_t i = v.size(); i > 0; i--) {
+      std::optional<int64_t> maybe = getConstantIntValue(v[i - 1]);
+      if (!maybe.has_value() || maybe.value() == 0) {
+        return i - 1;
+      }
+    }
+    return 0;
+  }
+
+  /// Get the highest index where the stride is 0. If this index is greater
+  /// than the lowest index where the size is greater than 1, then fail.
+  std::optional<uint32_t> verifyAndGetZeroStrideIndex(
+      ArrayRef<OpFoldResult> sizes, ArrayRef<OpFoldResult> strides,
+      NpuDmaCpyNdOp dmaOp) {
+    assert(strides.size() == sizes.size() && strides.size() == nbDimensions);
+
+    size_t firstNonUnitDim = getLowestIndexMaybeAboveOne(sizes);
+    size_t lastZeroStrideDim = getHighestIndexMaybeZero(strides);
+
+    if (firstNonUnitDim < lastZeroStrideDim) {
+      // Limitation until AIE-4.
```

Comment:
```suggestion
      // HW limitation.
```
I would drop AIE-4 mentions